### ISSUE TITLE
dockerd vulnerabilities

### ISSUE THREAD

**Issue by raghu017** [moby/moby#47246]
### Description

dockerd has several vulnerabilities. Can you fix them please?

https://hub.docker.com/layers/library/docker/25.0.1-dind/images/sha256-47d0ff54a405396b7ee5e882e33ee42d6457936b3393f2e40504cf03c0eb55ea?context=explore

[CVE-2023-47108](https://github.com/advisories/GHSA-8pgv-569h-w5rw) -
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc
g.org/grpc/otelgrpc

[CVE-2023-45142](https://github.com/advisories/GHSA-rcjv-mgp8-qvmr) -
go.opentelemetry.io/contrib/instrumentation/github.com/emicklei/go-restful/otelrestful
go.opentelemetry.io/contrib/instrumentation/github.com/gin-gonic/gin/otelgin
go.opentelemetry.io/contrib/instrumentation/github.com/gorilla/mux/otelmux
go.opentelemetry.io/contrib/instrumentation/github.com/labstack/echo/otelecho
go.opentelemetry.io/contrib/instrumentation/gopkg.in/macaron.v1/otelmacaron
go.opentelemetry.io/contrib/instrumentation/net/http/httptrace/otelhttptrace
go.opentelemetry.io/contrib/instrumentation/net/http/otel

[CVE-2023-48795](https://github.com/advisories/GHSA-45x7-px36-x8w8) -
golang.org/x/crypto

[GHSA-7ww5-4wqc-m92c](https://github.com/advisories/GHSA-7ww5-4wqc-m92c) -
github.com/containerd/containerd

### Reproduce

Remove vulnerable images

### docker version

```bash
24.0.8-dind, 25.0.1-dind
```

### docker info

```bash
24.0.8-dind, 25.0.1-dind
```

**Comment by thaJeztah** [comment]
Thanks for reporting; some of these CVEs look to be false positives, and others affect privileged APIs which, while potentially affected, would already provide `root` access (or equivalent) on the host, so would require a malicious actor to already have compromised the host. While (some of) these dependencies may be updated, we need to evaluate if the risk of updating outweighs the vulnerabilities from that perspective (i.e., updating the dependencies may impose a bigger risk of introducing regressions than the vulnerability that's reported).

These vulnerabilities were found in the 

- [CVE-2023-47108](https://github.com/advisories/GHSA-8pgv-569h-w5rw) affects OTEL tracing through the BuildKit and containerd gRPC APIs. Both APIs are considered privileged interfaces, and would require either access to the Docker Engine remote API (which is [equivalent to having `root` access on the host](https://docs.docker.com/engine/security/#docker-daemon-attack-surface)), or having root access on the host to access the `/var/run/containerd/containerd.sock` gRPC socket.  
- [CVE-2023-45142](https://github.com/advisories/GHSA-rcjv-mgp8-qvmr) requires access to the Engine Remote API using a TCP connection. This (as outlined above) is a privileged interface, equivalent to `root` access to the host.
 

These vulnerabilities were discovered in the `docker-buildx` and `containerd` binaries (which are not maintained in this repository, but in https://github.com/docker/buildx, and https://github.com/containerd/containerd);

- [GHSA-7ww5-4wqc-m92c](https://github.com/advisories/GHSA-7ww5-4wqc-m92c) is a false positive; the package is found in the `docker-buildx` CLI plugin, which uses the `github.com/containerd/containerd v1.7.7` dependency, but does not use the code related to the advisory.
- [CVE-2023-48795](https://github.com/advisories/GHSA-45x7-px36-x8w8) / [GO-2023-2402](https://pkg.go.dev/vuln/GO-2023-2402) is detected in the `docker-buildx` CLI plugin, which uses `golang.org/x/crypto v0.14.0` would potentially apply to connections made to the BuildKit gRPC API, which (outlined above) is considered a privileged API.
- [CVE-2023-44487](https://github.com/advisories/GHSA-m425-mq94-257g) is detected in the `docker-buildx` CLI plugin, which uses `google.golang.org/grpc 1.53.0`. This CVE also applies to connections made to the BuildKit gRPC API, which (outlined above) is considered a privileged API.

govulncheck results;


dockerd v25.0.1

<details>

```
git checkout v25.0.1

govulncheck ./...
Scanning your code and 1181 packages across 195 dependent modules for known vulnerabilities...

No vulnerabilities found.

Share feedback at https://go.dev/s/govulncheck-feedback.
```

</details>



docker v25.0.1 (docker cli):

<details>

```
git checkout v25.0.1

govulncheck ./...
Scanning your code and 443 packages across 71 dependent modules for known vulnerabilities...

No vulnerabilities found.

Share feedback at https://go.dev/s/govulncheck-feedback.
```

</details>



buildx v0.12.1:


<details>

```
git checkout v0.12.1

govulncheck ./...
Scanning your code and 1020 packages across 148 dependent modules for known vulnerabilities...

Vulnerability #1: GO-2023-2402
    Man-in-the-middle attacker can compromise integrity of secure channel in
    golang.org/x/crypto
  More info: https://pkg.go.dev/vuln/GO-2023-2402
  Module: golang.org/x/crypto
    Found in: golang.org/x/crypto@v0.14.0
    Fixed in: golang.org/x/crypto@v0.17.0
    Example traces found:
      #1: bake/remote.go:28:37: bake.ReadRemoteFiles calls dockerui.DetectGitContext, which eventually calls ssh.Dial

Vulnerability #2: GO-2023-2153
    Denial of service from HTTP/2 Rapid Reset in google.golang.org/grpc
  More info: https://pkg.go.dev/vuln/GO-2023-2153
  Module: google.golang.org/grpc
    Found in: google.golang.org/grpc@v1.53.0
    Fixed in: google.golang.org/grpc@v1.56.3
    Example traces found:
      #1: bake/remote.go:62:18: bake.ReadRemoteFiles calls client.Client.Build, which eventually calls grpc.NewServer
      #2: controller/remote/controller.go:180:24: remote.serveCmd calls grpc.Server.Serve, which eventually calls transport.NewServerTransport
      #3: controller/remote/controller.go:180:24: remote.serveCmd calls grpc.Server.Serve

=== Informational ===

Found 1 vulnerability in packages that you import, but there are no
call stacks leading to the use of this vulnerability. There is also 1
vulnerability in modules that you require that is neither imported nor
called. You may not need to take any action.
See https://pkg.go.dev/golang.org/x/vuln/cmd/govulncheck for details.

Vulnerability #1: GO-2023-2412
    RAPL accessibility in github.com/containerd/containerd
  More info: https://pkg.go.dev/vuln/GO-2023-2412
  Module: github.com/containerd/containerd
    Found in: github.com/containerd/containerd@v1.7.7
    Fixed in: github.com/containerd/containerd@v1.7.11

Vulnerability #2: GO-2023-2113
    Memory exhaustion in go.opentelemetry.io/contrib/instrumentation
  More info: https://pkg.go.dev/vuln/GO-2023-2113
  Module: go.opentelemetry.io/contrib/instrumentation/net/http/httptrace/otelhttptrace
    Found in: go.opentelemetry.io/contrib/instrumentation/net/http/httptrace/otelhttptrace@v0.40.0
    Fixed in: go.opentelemetry.io/contrib/instrumentation/net/http/httptrace/otelhttptrace@v0.44.0

  Module: go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp
    Found in: go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.40.0
    Fixed in: go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.44.0

Your code is affected by 2 vulnerabilities from 2 modules.

Share feedback at https://go.dev/s/govulncheck-feedback.
```

</details>

containerd v1.7.12

<details>

```
git checkout v1.7.12

govulncheck ./...
Scanning your code and 649 packages across 131 dependent modules for known vulnerabilities...

=== Informational ===

There is 1 vulnerability in modules that you require that is neither
imported nor called. You may not need to take any action.
See https://pkg.go.dev/golang.org/x/vuln/cmd/govulncheck for details.

Vulnerability #1: GO-2023-2402
    Man-in-the-middle attacker can compromise integrity of secure channel in
    golang.org/x/crypto
  More info: https://pkg.go.dev/vuln/GO-2023-2402
  Module: golang.org/x/crypto
    Found in: golang.org/x/crypto@v0.14.0
    Fixed in: golang.org/x/crypto@v0.17.0

No vulnerabilities found.

Share feedback at https://go.dev/s/govulncheck-feedback.

```

</details>

**Comment by raghu017** [comment]
That's good. When can I expect a fix for the high severity vulnerabilities mentioned earlier?

**Comment by raghu017** [comment]
Looking for vulnerability fix in latest images. 

Otelhttp is wip https://github.com/moby/moby/pull/47245 when would this be done?

> [CVE-2023-48795](https://github.com/advisories/GHSA-45x7-px36-x8w8) / [GO-2023-2402](https://pkg.go.dev/vuln/GO-2023-2402) is detected in the docker-buildx CLI plugin, which uses golang.org/x/crypto v0.14.0 would potentially apply to connections made to the BuildKit gRPC API, which (outlined above) is considered a privileged API.

Are you sure its buildx? They are on v0.17.0 for more than a month now - https://github.com/docker/buildx/blob/0326d2a5b1a9c13a53206db9ed871f23c71eacef/go.mod#L156

**Comment by thaJeztah** [comment]
That's the master branch (v0.13-dev, unreleased) current buildx releases are built from the v0.12 release branch; https://github.com/docker/buildx/blob/v0.12/go.mod#L42

**Comment by raghu017** [comment]
Any update on resolving the Otel vulnerabilities?
[CVE-2023-47108](https://github.com/advisories/GHSA-8pgv-569h-w5rw)

[CVE-2023-45142](https://github.com/advisories/GHSA-rcjv-mgp8-qvmr)

**Comment by raghu017** [comment]
@thaJeztah any update on resolving the Otel vulnerabilities?

**Comment by raghu017** [comment]
@thaJeztah Can I get a release for the Otel vulnerability fix https://github.com/moby/buildkit/pull/4677?

**Comment by thaJeztah** [comment]
It's merged into master, and will be in the v0.13 release of BuildKit, which part of the v26.0 release of docker engine.

However, as described above, those vulnerabilities require a malicious actor to have access to the Engine API, which already provides full root access to the host. Is there a scenario you have which would not fall into that category? Or is this only because vulnerability scanners flag the code?

**Comment by raghu017** [comment]
Sorry I meant this https://github.com/moby/moby/pull/47245. When can I get a release for this?

We are required to have zero vulnerabilities. If you think there is no vulnerability you should update your scanner or the scanner tool you are using to not show the vulnerability here https://hub.docker.com/layers/library/docker/dind/images/sha256-826fc608d01af1f6fba86ee2df8332d10d7c0a3361d95a02dbd20d2c122496c8?context=explore

**Comment by raghu017** [comment]
> It's merged into master, and will be in the v0.13 release of BuildKit, which part of the v26.0 release of docker engine.
> 
> However, as described above, those vulnerabilities require a malicious actor to have access to the Engine API, which already provides full root access to the host. Is there a scenario you have which would not fall into that category? Or is this only because vulnerability scanners flag the code?

Sorry I meant this https://github.com/moby/moby/pull/47245. When can I get a release for this?

We are required to have zero vulnerabilities. If you think there is no vulnerability you should update your scanner or the scanner tool you are using to not show the vulnerability here https://hub.docker.com/layers/library/docker/dind/images/sha256-826fc608d01af1f6fba86ee2df8332d10d7c0a3361d95a02dbd20d2c122496c8?context=explore

**Comment by neersighted** [comment]
Hi,

Reiterating what was stated above, we do not consider this issue exploitable in the Moby project as a user with access to the affected gRPC socket already is considered fully privileged, and may conduct actions much worse than a denial of service.

Simply put, this is outside our security model.

I am sorry that your organizations policy mandates you take action on non-exploitable problems like this. However, there is no action this project can take to help you in the way you ask.

We take security seriously, but a close second pillar is stability and quality. That is to say, bumping a dependency in haste or anger (e.g. backporting this PR in a patch release) to resolve a false finding like this is not a risk we are willing to take as maintainers.

As has been already communicated, the next major release will include this PR. There is no timeline for the next major beyond 'imminent' at this time.

Please understand that we are working to resolve this issue, and we do understand your perspective. However as maintainers we have a responsibility to all of our users, not just you, and we must balance priorities and risk for the collective best interest of the project.

I hope you understand, and please know that we are happy to discuss other issues with you in the future.
